### ISSUE TITLE
Unify tag enums in script land

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#1771]
I ran into a problem working #1571 where the split between `Analyzer::Tag` and `PacketAnalyzer::Tag` was causing issues with sending events. Some conversation ensued and the topic of unifying the tag types came up. I'm copying and pasting all of the conversation from that issue here to carry on fixing this separate from that issue.

https://github.com/zeek/zeek/issues/1571#issuecomment-916897862 (@rsmmr):

> Longer-term I'm wondering if it would make sense to unity all our analyzer tags into a single type so that we have them all inside the same enum namespace. The current separation has been causing trouble and/or replication in a few areas already.

https://github.com/zeek/zeek/issues/1571#issuecomment-917069434 (@timwoj):

> I was thinking the same this morning. I almost wonder if it'd be a good idea to shelve the tunnel work and get this resolved first. It's already affecting multiple things in the tunnel work, with protocol_confirmation/violation and the BPF filtering relying on the tags. Fixing it would make the remainder of the work easier.

https://github.com/zeek/zeek/issues/1571#issuecomment-917313170 (@timwoj):

> 
> 
> > Longer-term I'm wondering if it would make sense to unity all our analyzer tags into a single type so that we have them all inside the same enum namespace. The current separation has been causing trouble and/or replication in a few areas already.
> 
> I took a stab at this in [ff8f329](https://github.com/zeek/zeek/commit/ff8f329c7a87f172758da68011d15f2dea93f097) and tested with the following script:
> 
> ```
> function testing(t : Analyzer::Tag) : bool
> {
> print t;
> print enum_to_int(t);
> return T;
> }
> 
> event zeek_init()
> {
> testing(PacketAnalyzer::ANALYZER_TEREDO);
> testing(Analyzer::ANALYZER_HTTP);
> }
> ```
> 
> that outputs
> 
> ```
> PacketAnalyzer::ANALYZER_TEREDO
> 87
> Analyzer::ANALYZER_HTTP
> 24
> ```
> 
> Passing `-B plugins` shows that all of the analyzers of both kinds get unique tag numbers as well.

https://github.com/zeek/zeek/issues/1571#issuecomment-917962244 (@rsmmr):

> 
> 
> Interesting, that seems easier than I thought! Couple questions:
> 
>     1. This changes only the C++ side, so we would then still have to make a similar change for scripts, right?
> 
>     2. This is changing the plugin API. Would it work to keep an alias type inside the packet analyzer namespace to keep it source code compatible?
> 
>     3. We have a similar issue with the file analyzers. It's not quite as pressing there (because they don't overlap in semantics the way packet/session does), but would be nice to integrate them as well if that's viable (not sure)?

https://github.com/zeek/zeek/issues/1571#issuecomment-918469996 (@timwoj):

> 
> 
> > This changes only the C++ side, so we would then still have to make a similar change for scripts, right?
> 
> Surprisingly, none of the scripts change. The namespace for PacketAnalyzer tags still exists in script land because of how `TaggedComponent` works. The enum value is stored/increased as a static number to the Tag class itself, so if all of the tags are using the same class underneath then they all get the same span of enum values. We create the components in `packet_analyzer::Manager` and pass `PacketAnalyzer` to the constructor so the namespace remains.
> 
> EDIT: I see what you might mean here since the entire type (`PacketAnalyzer::Tag`) still exists in script land too. That should probably change. It's hard because there's no actual type checking, so you can pass `Analyzer::Tag` entries to `PacketAnalyzer::Tag` entries and vice-versa, but because they're just enums they go through. Unfortunately, just changing `PacketAnalyzer` to `Analyzer` in the constructor for `packet_analyzer::Manager` causes a type clash at startup.
> 
> > This is changing the plugin API. Would it work to keep an alias type inside the packet analyzer namespace to keep it source code compatible?
> 
> Putting `packet_analyzer/Tag.h` back and replacing it with just a deprecated alias seems to work just fine.
> 
> > We have a similar issue with the file analyzers. It's not quite as pressing there (because they don't overlap in semantics the way packet/session does), but would be nice to integrate them as well if that's viable (not sure)?
> 
> Can't all of the tag types be unified? `input` and `logging` both have custom tag types as well.

https://github.com/zeek/zeek/issues/1571#issuecomment-918691234 (@timwoj):

> The other thing is that Analyzer::Tag still gets passed down to script functions like function name(atype: Analyzer::Tag) : string, which throws an error if the enum value isn't an analyzer.

**Comment by timwoj** [comment]
My current thinking is somehow just unifying all of the tag types into one type, but adding some sort of enum to it so that we can determine what "type" of tag it is. From there we could add a new `TagManager` class that switches between the types so we can look things up via ``analyzer_mgr`,` `packet_mgr`, etc.

**Comment by rsmmr** [comment]
> > EDIT: I see what you might mean here since the entire type (`PacketAnalyzer::Tag`) still exists in script land too. That should probably change. It's hard because there's no actual type checking, so you can pass `Analyzer::Tag` entries to `PacketAnalyzer::Tag` entries and vice-versa, but because they're just enums they go through.

Yeah, that's what I meant. In the end, we'll need to change all events/funcitons that are taking one of the current tag types, to use the newly unified type instead. I was in particular thinking from the perspective of `protocol_confirmation` that triggered this all: we shouldn't be passing in an `PacketAnalyzer::Tag` there under the hood (even if it happens to work).

At the same time, it would be nice to keep alias types that redirect the old names (e.g., `PacketAnalyzer::Tag`) to the new type, so that existing scripts keep working. I don't remember if we can actually do that kind of aliasing  for script types?

> > Putting `packet_analyzer/Tag.h` back and replacing it with just a deprecated alias seems to work just fine.

Cool.

> > > We have a similar issue with the file analyzers. It's not quite as pressing there (because they don't overlap in semantics the way packet/session does), but would be nice to integrate them as well if that's viable (not sure)?
> > 
> > 
> > Can't all of the tag types be unified? `input` and `logging` both have custom tag types as well.

Yes, true, just didn't think of the others.

> > The other thing is that Analyzer::Tag still gets passed down to script functions like function name(atype: Analyzer::Tag) : string, which throws an error if the enum value isn't an analyzer.

Yeah, we can reimplement those I guess.

**Comment by rsmmr** [comment]
> My current thinking is somehow just unifying all of the tag types into one type, but adding some sort of enum to it so that we can determine what "type" of tag it is. 

Yeah, tags already have `type`/`subtype`, can add an enum to that (maybe call it `namespace`?)

> From there we could add a new `TagManager` class that switches between the types so we can look things up via `analyzer_mgr`, `packet_mgr`, etc.

Not sure what "switches between the types" means, but generally yes, agree.

Actually I like this line of work, will simplify stuff quite a bit.

**Comment by timwoj** [comment]
> I don't remember if we can actually do that kind of aliasing for script types?

Not that I'm aware of. @vpax do you know?

> The other thing is that Analyzer::Tag still gets passed down to script functions like function name(atype: Analyzer::Tag) : string, which throws an error if the enum value isn't an analyzer.

This throws an error because it tries to call `Analyzer::name()` on a packet analyzer tag, which then tries to look that tag up via `analyzer_mgr` to get the analyzer's name. If we were to store a name or something in the tag itself or in a tag manager, that could be avoided.

**Comment by vpax** [comment]
If I'm following the above correctly, then I believe something like `type PacketAnalyzer::Tag: Analyzer::Tag;` would introduce a working alias.

**Comment by timwoj** [comment]
I have the C++ side of this mostly done at this point. It's down to one tag type, and I removed some templating that was required by the split tag types. Now to figure out the script land side.

**Comment by timwoj** [comment]
Here’s where I am design-/code-wise at this point:

- All of the individual `Tag` classes in the various components are combined into one class now, with deprecated versions left behind for the old types.
- All uses of `Tag` in templates have been removed. This simplified `Component` down to not need a template at all, and `ComponentManager` only needs to take the Component class now instead of both a `Component` and a `Tag`.
- A global `EnumType` now exists in `Tag` that gets used for all of the places that individual tag types were stored by `ComponentManager`. This allowed me to combine all of the `Tag` enum values in script land into a single type in the global module. This value is a public static variable in `Tag` and is initialized during `setup()`.
- All of the existing `Tag` script-land types have a deprecation warning when using them, set during creation in `ComponentManager`.
- `ComponentManager::GetTagType()` is now deprecated, since 

My primary concerns right now:
- Combining all of the tag types might have some odd effects, like allowing you to pass an analyzer when you meant to pass a logger. Previously the script parser would have reported an error for that. It’s making me wonder if instead of one single global `EnumType`, it should be one for all of the analyzers (combining `Analyzer`, `PacketAnalyzer`, and `Files` into one), one for the logger framework, and one for the input framework. This would complicate the code on the C++ side a little bit, but not significantly.
- I’m still not sure how to deprecate the events. Just replacing them directly breaks things (I tried that). Will just adding a second version of each event that uses a tag type, and marking the other ones as deprecated work? Does this break something in third-party code?
